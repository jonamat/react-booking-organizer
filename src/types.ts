import { WEEK_DAYS } from './config';

/* -------------------------------------------------------------------------- */
/*                                    Utils                                   */
/* -------------------------------------------------------------------------- */

export type HourMinute = {
    hour: number;
    minute: number;
};

export type WorkingShift = {
    startTime: HourMinute;
    endTime: HourMinute;
};

export type WeekDays = 'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday' | 'sunday';
export type WeekWorkingShifts = {
    [key in WeekDays]?: Array<WorkingShift>;
};

export interface Holiday {
    startDay: Date;
    endDay: Date;
}

/* -------------------------------------------------------------------------- */
/*                                Reservations                                */
/* -------------------------------------------------------------------------- */

export interface Reservation {
    id: string;
    date: Date;
    serviceId: string;
    customerId: string;
    employeeId: string;
    notes?: string;
}
export type NewReservation = Omit<Reservation, 'id'>;
export function isNewReservation(obj: any): obj is NewReservation {
    if (typeof obj !== 'object' || obj === null) return false;
    return (
        Object.keys(obj).every((key) => {
            if (obj[key] === undefined) return true;
            switch (key) {
                case 'date':
                    return obj[key] instanceof Date;
                case 'serviceId':
                case 'customerId':
                case 'employeeId':
                case 'notes':
                    return typeof obj[key] === 'string';
                default:
                    return false;
            }
        }) && ['date', 'serviceId', 'customerId', 'employeeId'].every((required) => !!obj[required])
    );
}

/* -------------------------------------------------------------------------- */
/*                                  Customers                                 */
/* -------------------------------------------------------------------------- */

export interface Customer {
    id: string;
    name: string;
    surname?: string;
    additionalIdentifier?: string;
    phone?: string;
    birthday?: Date;
    notes?: string;
}
export type NewCustomer = Omit<Customer, 'id'>;

export function isNewCustomer(obj: any): obj is NewCustomer {
    if (typeof obj !== 'object' || obj === null) return false;
    return (
        Object.keys(obj).every((key) => {
            if (obj[key] === undefined) return true;
            switch (key) {
                case 'name':
                case 'surname':
                case 'additionalIdentifier':
                case 'phone':
                case 'notes':
                    return typeof obj[key] === 'string';
                case 'birthday':
                    return obj[key] instanceof Date;
                default:
                    return false;
            }
        }) && obj.name
    );
}

/* -------------------------------------------------------------------------- */
/*                                  Employees                                 */
/* -------------------------------------------------------------------------- */

export interface Employee {
    id: string;
    name: string;
    weekWorkingShifts?: WeekWorkingShifts;
    holidays?: Array<Holiday>;
}
export type NewEmployee = Omit<Employee, 'id'>;

export function isHoliday(obj: any): obj is Holiday {
    return (
        typeof obj === 'object' &&
        obj !== null &&
        Object.entries(obj).every(([key, date]) => ['startDay', 'endDay'].includes(key) && date instanceof Date)
    );
}

export function isWorkingShift(obj: any): obj is WorkingShift {
    if (typeof obj !== 'object' || obj === null) return false;
    return Object.entries(obj).every(
        ([key, hourMinute]) =>
            ['startTime', 'endTime'].includes(key) &&
            Object.entries(hourMinute as HourMinute).every(
                ([key, value]) => ['hour', 'minute'].includes(key) && typeof value === 'number',
            ),
    );
}

export function isWeekWorkingShifts(obj: any): obj is WeekWorkingShifts {
    if (typeof obj !== 'object' || obj === null) return false;
    return Object.entries(obj).every(([key, workingShifts]) => {
        if (!WEEK_DAYS.includes(key as WeekDays)) return false;
        if (typeof workingShifts === 'undefined') return true;
        return Array.isArray(workingShifts) && workingShifts.every((workingShift) => isWorkingShift(workingShift));
    });
}

export function isNewEmployee(obj: any): obj is NewEmployee {
    if (typeof obj !== 'object' || obj === null) return false;
    return (
        Object.keys(obj).every((key) => {
            if (obj[key] === undefined) return true;
            switch (key) {
                case 'name':
                    return typeof obj[key] === 'string';
                case 'weekWorkingShifts':
                    return obj.weekWorkingShifts === null || isWeekWorkingShifts(obj.weekWorkingShifts);
                case 'holidays':
                    return (
                        obj.holidays === null ||
                        (Array.isArray(obj.holidays) && obj.holidays.every((holiday) => isHoliday(holiday)))
                    );
                default:
                    return false;
            }
        }) && obj.name
    );
}

/* -------------------------------------------------------------------------- */
/*                                  Services                                  */
/* -------------------------------------------------------------------------- */

export interface Service {
    id: string;
    name: string;
    averageDuration: number; // hourly brackets
}
export type NewService = Omit<Service, 'id'>;

export function isNewService(obj: any): obj is NewService {
    if (typeof obj !== 'object' || obj === null) return false;
    return (
        Object.keys(obj).every((key) => {
            if (obj[key] === undefined) return true;
            switch (key) {
                case 'name':
                    return typeof obj[key] === 'string';
                case 'averageDuration':
                    return typeof obj[key] === 'number' && !isNaN(obj[key]);
                default:
                    return false;
            }
        }) && obj.name
    );
}

/* -------------------------------------------------------------------------- */
/*                                 REDUX TYPES                                */
/* -------------------------------------------------------------------------- */

/* ---------------------------------- Utils --------------------------------- */
export type Thunk<A> = {
    (...args: Array<any>): {
        (dispatch: (action: A | ((...args: any) => void)) => void, getState: () => RootState): void;
    };
};

export interface DialogStatus {
    open: boolean;
    title: string | null;
    content: JSX.Element | null;
}

/* -------------------------------------------------------------------------- */
/*                                  Database                                  */
/* -------------------------------------------------------------------------- */

/* -------------------------------- Reducers -------------------------------- */
export interface DatabaseReducer {
    customers: Array<Customer> | null;
    reservations: Array<Reservation> | null;
    employees: Array<Employee> | null;
    services: Array<Service> | null;
}

/* --------------------------------- Actions -------------------------------- */
export interface ResetDatabaseAction {
    readonly type: 'RESET_DATA';
}
export interface UpdateCustomersAction {
    readonly type: 'UPDATE_CUSTOMERS';
    customers: Array<Customer>;
}
export interface UpdateReservationsAction {
    readonly type: 'UPDATE_RESERVATIONS';
    reservations: Array<Reservation>;
}
export interface UpdateEmployeesAction {
    readonly type: 'UPDATE_EMPLOYEES';
    employees: Array<Employee>;
}
export type UpdateServicesAction = {
    readonly type: 'UPDATE_SERVICES';
    services: Array<Service>;
};

/* ----------------------------- Action creators ---------------------------- */
export type UpdateCustomersCreator = Thunk<UpdateCustomersAction>;
export type UpdateReservationsCreator = Thunk<UpdateReservationsAction>;
export type UpdateEmployeesCreator = Thunk<UpdateEmployeesAction>;
export type UpdateServicesCreator = Thunk<UpdateServicesAction>;

/* -------------------------- Actions intersection -------------------------- */
export type DatabaseActionTypes =
    | ResetDatabaseAction
    | UpdateCustomersAction
    | UpdateReservationsAction
    | UpdateEmployeesAction
    | UpdateServicesAction;

/* -------------------------------------------------------------------------- */
/*                                   Status                                   */
/* -------------------------------------------------------------------------- */

/* -------------------------------- Reducers -------------------------------- */
export interface StatusReducer {
    offline: boolean;
    dialog: DialogStatus;
    employeesOrder: null | Array<string>;
    firebaseSync: boolean;
    birthdaysToday: number;
    birthdayNotificationWasDisplayed: boolean;
}

/* --------------------------------- Actions -------------------------------- */
export interface CheckSyncAction {
    readonly type: 'FIREBASE_SYNCHRONIZED';
    firebaseSync: boolean;
}
export interface IsOfflineAction {
    readonly type: 'IS_OFFLINE';
    isOffline: boolean;
}
export interface DialogAction {
    readonly type: 'CHANGE_DIALOG_STATUS';
    dialog: DialogStatus;
}
export interface MoveEmployeeAction {
    readonly type: 'UPDATE_EMPLOYEES_ORDER';
    employeesOrder: Array<string>;
}
export interface UpdateBirthdaysTodayAction {
    readonly type: 'UPDATE_BIRTHDAYS_TODAY';
    birthdaysToday: number;
}
export interface BirthdaysNotificationDisplayedAction {
    readonly type: 'BIRTHDAYS_NOTIFICATION_DISPLAYED';
    birthdayNotificationWasDisplayed: boolean;
}

/* ----------------------------- Action creators ---------------------------- */
export type IsOfflineCreator = Thunk<IsOfflineAction>;
export type DialogCreator = Thunk<DialogAction>;
export type MoveEmployeeCreator = Thunk<MoveEmployeeAction>;
export type CheckSyncCreator = Thunk<CheckSyncAction>;
export type UpdateBirthdaysTodayCreator = Thunk<UpdateBirthdaysTodayAction>;
export type BirthdaysNotificationDisplayedCreator = Thunk<BirthdaysNotificationDisplayedAction>;

/* -------------------------- Actions intersection -------------------------- */
export type StatusActionTypes =
    | IsOfflineAction
    | DialogAction
    | MoveEmployeeAction
    | CheckSyncAction
    | UpdateBirthdaysTodayAction
    | BirthdaysNotificationDisplayedAction;

/* -------------------------------------------------------------------------- */
/*                                    Auth                                    */
/* -------------------------------------------------------------------------- */

/* -------------------------------- Reducers -------------------------------- */
export interface AuthReducer {
    logged: boolean | null;
}

/* --------------------------------- Actions -------------------------------- */
export interface IsLoggedAction {
    readonly type: 'IS_LOGGED';
    isLogged: boolean;
}

/* ----------------------------- Action creators ---------------------------- */
export type IsLoggedCreator = Thunk<IsLoggedAction>;

/* -------------------------- Actions intersection -------------------------- */
export type AuthActionTypes = IsLoggedAction;

/* -------------------------------------------------------------------------- */
/*                                  RootState                                 */
/* -------------------------------------------------------------------------- */

export type RootState = {
    readonly auth: AuthReducer;
    readonly status: StatusReducer;
    readonly database: DatabaseReducer;
};
