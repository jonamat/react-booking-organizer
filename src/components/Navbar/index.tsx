import React, { FC, ReactNode } from 'react';
import { RouteComponentProps, withRouter } from 'react-router';
import { connect, ConnectedProps } from 'react-redux';
import { Tooltip, SvgIcon, AppBar, Toolbar, IconButton, Typography, TextField, Badge } from '@material-ui/core';
import Autocomplete from '@material-ui/lab/Autocomplete';

import { Customer, RootState } from '../../types';
import { getReservationsByDate } from '../../utils';
import Logo from '../../assets/img/logo_white.svg';
import useStyles from './style';

// Icons
import AssignmentIcon from '@material-ui/icons/Assignment';
import SearchIcon from '@material-ui/icons/Search';

const mapStateToProps = (state: RootState) => ({
    reservations: state.database.reservations,
    customers: state.database.customers,
});
const connector = connect(mapStateToProps);
type ReduxProps = ConnectedProps<typeof connector>;

interface NavbarProps extends RouteComponentProps, ReduxProps {
    toggleDrawer: () => void;
}

const Navbar: FC<NavbarProps> = ({ customers, reservations, history, toggleDrawer }) => {
    const classes = useStyles();

    if (!reservations || !customers) {
        console.warn('Internal error - Resources loading logic broken');
        reservations = reservations || [];
        customers = customers || [];
    }

    const reservationsToday = getReservationsByDate(reservations, new Date()).length;

    const handleCustomerSearch = (customer: Customer | null) => {
        customer && history.push('/customer-details/' + customer.id);
    };

    return (
        <div className={classes.grow} data-testid="navbar">
            <AppBar position="sticky">
                <Toolbar classes={{ gutters: classes.navbarGutters }}>
                    <IconButton
                        className={classes.drawerButton}
                        edge="start"
                        onClick={toggleDrawer}
                        color="inherit"
                        data-testid="drawer-btn"
                    >
                        <SvgIcon>
                            <Logo />
                        </SvgIcon>
                    </IconButton>

                    <SvgIcon classes={{ root: classes.logo }}>
                        <Logo />
                    </SvgIcon>

                    <Typography className={classes.title} variant="h6" noWrap>
                        Gestionale {process.env.COMPANY_NAME.toUpperCase()}
                    </Typography>
                    <div className={classes.grow} />
                    <div className={classes.search}>
                        <Autocomplete
                            classes={{
                                popupIndicator: classes.autocompleteIndicators,
                                clearIndicator: classes.autocompleteIndicators,
                            }}
                            options={customers}
                            getOptionLabel={(customer: Customer): string =>
                                customer.name +
                                (customer.surname ? ' ' + customer.surname : '') +
                                (customer.additionalIdentifier ? ' ' + customer.additionalIdentifier : '')
                            }
                            autoHighlight
                            /**
                             * This actually does not work due to a Autocomplete bug
                             * https://github.com/mui-org/material-ui/issues/20286
                             * use onMouseDownCapture workaround
                             */
                            openOnFocus={false}
                            clearText="Cancella"
                            closeText="Chiudi"
                            noOptionsText="Cliente non trovato"
                            onChange={(_event: any, customer: Customer | null) => handleCustomerSearch(customer)}
                            renderInput={(params): ReactNode => (
                                <TextField
                                    {...params}
                                    classes={{
                                        root: classes.textField,
                                    }}
                                    InputProps={{
                                        ...params.InputProps,
                                        disableUnderline: true,
                                        startAdornment: <SearchIcon />,
                                        classes: {
                                            root: classes.input,
                                        },
                                        onMouseDownCapture: (e) => e.stopPropagation(),
                                    }}
                                    inputProps={{
                                        ...params.inputProps,
                                        'data-testid': 'search-field',
                                        id: 'Autocomplete',
                                    }}
                                    placeholder="Cercaâ€¦"
                                />
                            )}
                        />
                    </div>
                    <Tooltip title="Appuntamenti oggi">
                        <IconButton color="inherit" onClick={() => history.push('/')}>
                            <Badge badgeContent={reservationsToday} color="secondary">
                                <AssignmentIcon data-testid="today-btn" />
                            </Badge>
                        </IconButton>
                    </Tooltip>
                </Toolbar>
            </AppBar>
        </div>
    );
};

export default connector(withRouter(Navbar));
