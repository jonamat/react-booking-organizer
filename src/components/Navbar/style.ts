import { makeStyles, Theme, fade } from '@material-ui/core';

const useStyles = makeStyles((theme: Theme) => ({
    grow: {
        flexGrow: 2,
    },
    navbarGutters: {
        padding: theme.spacing(0, 1, 0, 2),
    },
    drawerButton: {
        display: 'block',
        marginRight: theme.spacing(2),
        [theme.breakpoints.up('sm')]: {
            display: 'none',
        },
    },
    logo: {
        display: 'none',
        width: '1.5em',
        height: '1.5em',
        marginRight: theme.spacing(3),
        [theme.breakpoints.up('sm')]: {
            display: 'block',
        },
    },
    title: {
        display: 'none',
        fontFamily: '"EB Garamond", serif',
        fontSize: 25,
        letterSpacing: '0.03em',
        [theme.breakpoints.up('sm')]: {
            display: 'block',
        },
    },
    search: {
        flexGrow: 1,
        padding: theme.spacing(0, 2),
        marginRight: theme.spacing(1),
        borderRadius: theme.shape.borderRadius,
        backgroundColor: fade(theme.palette.common.white, 0.3),
        transition: theme.transitions.create(['opacity', 'width']),
        opacity: 0.7,
        '&:hover, &:focus-within': {
            backgroundColor: fade(theme.palette.common.white, 0.3),
            opacity: 1,
        },
        width: '100%',
        [theme.breakpoints.up('sm')]: {
            marginLeft: theme.spacing(3),
            width: 80,
            '&:focus-within': {
                width: 250,
            },
        },
    },
    autocompleteIndicators: {
        color: 'inherit',
    },
    textField: {
        width: '100%',
        margin: theme.spacing(1, 0, 1, 0),
    },
    input: {
        flexWrap: 'nowrap',
        color: 'inherit',
        width: '100%',
    },
}));

export default useStyles;
