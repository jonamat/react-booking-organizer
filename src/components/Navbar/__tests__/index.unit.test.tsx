/**
 * Unit test suite
 *
 * Target: Navbar component
 * Tests:
 * - Autocomplete integration
 * - Snapshot matching
 * - Functionality
 */

// Test resources
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { createMemoryHistory } from 'history';
import { StylesProvider, createGenerateClassName } from '@material-ui/core';

// Test configuration
import { Customer, Reservation } from '../../../types';

// Component to test
import Navbar from '../';

const customers: Array<Customer> = [
    {
        id: 'customer-FAKE-ID',
        name: 'customer-FAKE-NAME',
        additionalIdentifier: 'customer-FAKE-IDENTIFIER',
    },
];
const mockReservation: Reservation = {
    id: 'reservation-FAKE-ID',
    date: new Date(),
    serviceId: 'service-FAKE-ID',
    customerId: 'customer-FAKE-ID',
    employeeId: 'employee-FAKE-ID',
    notes: 'FAKE-NOTES',
};
const reservations: Array<Reservation> = [mockReservation];

/* -------------------------------------------------------------------------- */
/*                                    Mocks                                   */
/* -------------------------------------------------------------------------- */

jest.mock('react-router', () => ({
    withRouter: (component) => component,
}));
jest.mock('react-redux', () => ({
    connect: () => (component) => component,
}));
jest.mock('../../../utils', () => ({
    getReservationsByDate: function (): Array<Reservation> {
        return [mockReservation];
    },
}));
const history = {
    push: jest.fn(),
};
const mockToggleDrawer = jest.fn();

/* -------------------------------------------------------------------------- */
/*                              Bridge operations                             */
/* -------------------------------------------------------------------------- */

beforeEach(() => {
    jest.clearAllMocks();
});

/* -------------------------------------------------------------------------- */
/*                               Render wrapper                               */
/* -------------------------------------------------------------------------- */

/**
 * Render the component with the custom configuration
 * @param props Component props
 */
const renderComponent = (props?) => {
    return render(
        <Navbar
            history={history}
            customers={customers}
            reservations={reservations}
            toggleDrawer={mockToggleDrawer}
            {...props}
        />,
    );
};

/* -------------------------------------------------------------------------- */
/*                                    Tests                                   */
/* -------------------------------------------------------------------------- */

describe('Navbar', () => {
    describe('Autocomplete', () => {
        it('sorts customers properly', () => {
            const { getByTestId } = renderComponent();
            fireEvent.click(getByTestId('search-field'));
            fireEvent.input(getByTestId('search-field'), { target: { value: 'customer' } });
            fireEvent.keyDown(getByTestId('search-field'), { key: 'Enter', code: 'Enter' });
            expect(getByTestId('search-field')).toHaveValue('customer-FAKE-NAME customer-FAKE-IDENTIFIER');
        });

        it('redirects to customer-details page when a customer is triggered from list', () => {
            const { getByTestId } = renderComponent();
            fireEvent.click(getByTestId('search-field'));
            fireEvent.input(getByTestId('search-field'), { target: { value: 'customer' } });
            fireEvent.keyDown(getByTestId('search-field'), { key: 'Enter', code: 'Enter' });
            expect(history.push).toBeCalledWith('/customer-details/customer-FAKE-ID');
        });
    });

    describe('Today button', () => {
        it('redirects to Today page on Today button click', () => {
            const { getByTestId } = renderComponent();
            fireEvent.click(getByTestId('today-btn'));
            expect(history.push).toBeCalledWith('/');
        });

        it('badge shows correct reservations number', () => {
            const { getByTestId } = renderComponent();
            expect(getByTestId('today-btn').nextSibling).toHaveTextContent('1');
        });
    });

    describe('Drawer button', () => {
        it('calls drawer at click', () => {
            const { getByTestId } = renderComponent();
            fireEvent.click(getByTestId('drawer-btn'));
            expect(mockToggleDrawer).toHaveBeenCalled();
        });
    });

    it('matches snapshot', () => {
        const history = createMemoryHistory();
        const snapshot = renderer
            .create(
                <StylesProvider generateClassName={createGenerateClassName({ disableGlobal: true })}>
                    {/* @ts-ignore */}
                    <Navbar history={history} customers={customers} reservations={reservations} />
                </StylesProvider>,
            )
            .toJSON();
        expect(snapshot).toMatchSnapshot();
    });
});
