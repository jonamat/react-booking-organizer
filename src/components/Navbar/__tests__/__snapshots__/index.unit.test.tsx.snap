// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar matches snapshot 1`] = `
<div
  className="jss1"
  data-testid="navbar"
>
  <header
    className="jss21 jss10 jss13 jss17 jss28"
  >
    <div
      className="jss49 jss51 jss50 jss2"
    >
      <button
        className="jss62 jss53 jss3 jss56 jss54"
        data-testid="drawer-btn"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="jss61"
        >
          <svg
            aria-hidden={true}
            className="jss65"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <test-file-stub />
          </svg>
        </span>
        <span
          className="jss191"
        />
      </button>
      <svg
        aria-hidden={true}
        className="jss65 jss4"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <test-file-stub />
      </svg>
      <h6
        className="jss74 jss5 jss84 jss93"
      >
        Gestionale 
        DEMO
      </h6>
      <div
        className="jss1"
      />
      <div
        className="jss6"
      >
        <div
          aria-expanded={false}
          aria-owns={null}
          className="jss104 jss110 jss109"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="combobox"
        >
          <div
            className="jss129 jss128 jss8 jss132"
          >
            <div
              className="jss147 jss133 jss9 jss111 jss157 jss142 jss148 jss134 jss151 jss152"
              onClick={[Function]}
              onMouseDownCapture={[Function]}
            >
              <svg
                aria-hidden={true}
                className="jss65"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-controls={null}
                aria-invalid={false}
                autoCapitalize="none"
                autoComplete="off"
                autoFocus={false}
                className="jss158 jss143 jss112 jss113 jss162 jss163"
                data-testid="search-field"
                disabled={false}
                id="Autocomplete"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                placeholder="Cercaâ€¦"
                required={false}
                spellCheck="false"
                type="text"
                value=""
              />
              <div
                className="jss114"
              >
                <button
                  aria-label="Cancella"
                  className="jss62 jss53 jss115 jss7"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  title="Cancella"
                  type="button"
                >
                  <span
                    className="jss61"
                  >
                    <svg
                      aria-hidden={true}
                      className="jss65 jss72"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    className="jss191"
                  />
                </button>
                <button
                  aria-label="Open"
                  className="jss62 jss53 jss117 jss7"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  title="Open"
                  type="button"
                >
                  <span
                    className="jss61"
                  >
                    <svg
                      aria-hidden={true}
                      className="jss65"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                  <span
                    className="jss191"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        aria-describedby={null}
        className="jss62 jss53 jss56"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Appuntamenti oggi"
        type="button"
      >
        <span
          className="jss61"
        >
          <span
            className="jss176"
          >
            <svg
              aria-hidden={true}
              className="jss65"
              data-testid="today-btn"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3h-4.18C14.4 1.84 13.3 1 12 1c-1.3 0-2.4.84-2.82 2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7 0c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zm2 14H7v-2h7v2zm3-4H7v-2h10v2zm0-4H7V7h10v2z"
              />
            </svg>
            <span
              className="jss177 jss182 jss179"
            >
              1
            </span>
          </span>
        </span>
        <span
          className="jss191"
        />
      </button>
    </div>
  </header>
</div>
`;
